### PR TITLE
[BUG] Fixed SARIMAX fail when X is passed to predict but not fit

### LINKED ISSUE #5789: [BUG] `SARIMAX` fails when `X` is passed in `predict` but not used in `fit`
## Minimal Reproducible Example

```pycon
>>> 
>>> from sktime.datasets import load_longley
>>> from sktime.forecasting.sarimax import SARIMAX
>>> from sktime.split import temporal_train_test_split
>>> 
>>> y, X = load_longley()
>>> 
>>> y_train, _, _, X_test = temporal_train_test_split(y, X)
>>> 
>>> forecaster = SARIMAX()
>>> 
>>> forecaster.fit(y_train)
SARIMAX()
>>> 
>>> # works
>>> forecaster.predict(fh=[1, 2, 3, 4])
1959    66061.176439
1960    65682.034815
1961    65363.883253
1962    65096.910677
Freq: A-DEC, Name: TOTEMP, dtype: float64
>>> 
>>> # fails
>>> forecaster.predict(fh=[1, 2, 3, 4], X=X_test)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/anirban/sktime-fork/sktime/forecasting/base/_base.py", line 412, in predict
    y_pred = self._predict(fh=fh, X=X_inner)
  File "/home/anirban/sktime-fork/sktime/forecasting/base/adapters/_statsmodels.py", line 108, in _predict
    ind_drop = self._X.index
AttributeError: 'NoneType' object has no attribute 'index'
>>> 
```

## Expectation

I was expecting no failures, and identical behaviour in both cases. I do get that behaviour if I try with `sktime.forecasting.arima.ARIMA`.

## Version

Operating System: Ubuntu 22.04.3 LTS (WSL)
Python: 3.10.12
Sktime: e51ec2472a

### PR DESCRIPTION
**Pull request by Abhay-Lejith**
#### Reference Issues/PRs

This fixes #5789 .

#### What does this implement/fix? Explain your changes.

`SARIMAX` no longer gives an error when `X` is passed to `predict` but not to `fit`. The `X` passed to `predict` is ignored in this case i.e. it is changed to `None`.
#### Does your contribution introduce a new dependency? If yes, which one?
No.

#### Did you add any tests for the change?
No.

#### Any other comments?
All check estimator tests have passed when I checked it locally. Do I have to add any additional tests? If yes, how do I go about it?

##### For all contributions
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Comment by yarnabrina** [comment]
This has a much bigger scope than just `SARIMAX`, as this is a change in the general adapter for all forecasters on top of statsmodels. FYI @ciaran-g as adapter maintainer.

Question to @ciaran-g / @fkiraly / @benHeid : what happens if user passes full `X` to `fit` call? Will it be indexed according to index of y and only the correct subset will be stored in `self._X` and passed to `_fit`, or will the full data be used? How is it affected by `X-y-must-have-same-index` tag?

File: sktime/forecasting/base/adapters/_statsmodels.py
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_statsmodels.py:121]: I'm not sure if I'm clear about when this check will evaluate to True. Can you please explain?

File: sktime/forecasting/base/adapters/_statsmodels.py
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_statsmodels.py:114]: Optional: may be merge with previous check with `and`?

File: sktime/forecasting/base/adapters/_statsmodels.py
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_statsmodels.py:115]: This is the question I asked yesterday. Is it guaranteed that `ind_drop` only contains indices from training range and never from prediction range?

**Review by yarnabrina — CHANGES_REQUESTED** [review]
I added some review comments, mostly optional and may be out of scope of this PR (in case if it's a different bug).

I'll request you to add a test case for SARIMAX at least. You can use the same example in the bhg report and check it works correctly after this PR.

File: sktime/forecasting/base/adapters/_statsmodels.py
Comment (Abhay-Lejith) [reply sktime/forecasting/base/adapters/_statsmodels.py:121]: This will evaluate to True when the user has not passed X to fit (hence self._X is None) but has passed X into predict.

File: sktime/forecasting/base/adapters/_statsmodels.py
Comment (Abhay-Lejith) [reply sktime/forecasting/base/adapters/_statsmodels.py:115]: Here `ind_drop` contains the indices present in the `X` passed to `fit` (which should be the training set ideally).

Comment (Abhay-Lejith) [reply sktime/forecasting/base/adapters/_statsmodels.py:115]: Here `ind_drop` contains the indices present in the `X` passed to `fit` (which should be the training set ideally).

File: sktime/forecasting/base/adapters/_statsmodels.py
Comment (Abhay-Lejith) [reply sktime/forecasting/base/adapters/_statsmodels.py:114]: I will check this out.

**Comment by Abhay-Lejith** [comment]
> I added some review comments, mostly optional and may be out of scope of this PR (in case if it's a different bug).
> 
> I'll request you to add a test case for SARIMAX at least. You can use the same example in the bhg report and check it works correctly after this PR.

How and where do I add the tests ? Please link me to some resource or any help would be great. Haven't done this before...

File: sktime/forecasting/base/adapters/_statsmodels.py
Comment (Abhay-Lejith) [reply sktime/forecasting/base/adapters/_statsmodels.py:114]: Can be done ,I'll make the change.

File: sktime/forecasting/base/adapters/_statsmodels.py
Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_statsmodels.py:115]: Ideally and more common, yes. Not sure if it's a guarantee. For example, I personally have definitely passed full `X` sometimes (not necessarily to SARIMAX, but I don't remember exactly).

File: sktime/forecasting/base/adapters/_statsmodels.py
Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_statsmodels.py:121]: I was thinking of another estimator, where "exog" will be present in the fitted forecaster depending on whether it's trained with X or not. I checked now with SARIMAX/VAR/VARMAX and that doesn't seemed to the case, and I can't remember where I saw that type of behaviour. Sorry for the confusion.

**Comment by yarnabrina** [comment]
> I just had a look and found a test_sarimax.py file. Do I add an extra function like these that are there already in the file?

Right, exactly. These are general steps for a testing function:

- you write a function that returns nothing (or None if you want)
- the function executes few lines of codes and check if some expected properties are correct or not
- expected properties can be a code snippet must fail with a specific error type and specific error message in a paritular case
- or that a particular code snippet must work without failure
- or that a particular code snippet must generate a result that matches a preset outcome which is known to be correct

In `sktime`, we have lots of soft dependencies which are not always installed. So to protect from import errors, we add the decorator function to check if all necessary package dependencies for a function is present. The pytest.skip call ensures that the test does not run (and eventually fail) if the requirements are not satisfied.

For this particular case, my suggestion is you create a new test function in `test_sarimax.py`, copy paste my original failing code, and add the decorator checking for dependencies of SARIMAX class. Then you can run the test locally with `pytest -k "<name of your test function>"` and see whether it passes now or not.

File: sktime/forecasting/base/adapters/_statsmodels.py
Comment (Abhay-Lejith) [reply sktime/forecasting/base/adapters/_statsmodels.py:115]: Ok so in the event that a user passes full `X` to SARIMAX in fit , what will they pass to predict? If it is a subset of this `X`, (say `test_X` obtained by splitting `X`) it will result in all it's indices being dropped and the code will run without error but the exog features will not be used in this case. However ideally the user should be passing different X to fit and predict in order for it to serve it's purpose.
This code to drop indices was added as a fix for issue #3830 as mentioned in the comment.
https://github.com/sktime/sktime/blob/14c88c69d2a0e961390e749278b5349470d35332/sktime/forecasting/base/adapters/_statsmodels.py#L110-L115
We technically cannot do anything about `full X` being passed to `fit` because we do not know what the user will pass to `predict` next.

Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_statsmodels.py:115]: > it will result in all it's indices being dropped and the code will run without error

Can you please validate this? My intinct says that the underlying model (from `statsmodels`) will complain if `X` is not passed during `predict` but passed during `fit`.

Comment (Abhay-Lejith) [reply sktime/forecasting/base/adapters/_statsmodels.py:115]: >it will result in all it's indices being dropped and the code will run without error

Small correction here, I meant this in the case where `full X` is passed to `fit` and a `subset of X` is passed to `predict`. However I was wrong in saying that code runs without error, rather all indices get dropped and it returns an `error` saying the received exog is not in the correct shape.
```python
from sktime.datasets import load_longley
from sktime.split import temporal_train_test_split
from sktime.forecasting.sarimax import SARIMAX

y, X = load_longley()
y_train, _, X_train, X_test = temporal_train_test_split(y, X)
forecaster = SARIMAX()
forecaster.fit(y,X=X)
forecaster.predict(fh=[1, 2, 3, 4],X=X_train)
```
```python
ValueError: Provided exogenous values are not of the appropriate shape. Required (4, 5), got (0, 5).
```
Also, if X is passed to `fit` but not `predict`, it returns the following error asking for exog features to be passed in predict as well.
```python
ValueError: Out-of-sample operations in a model with a regression component require additional exogenous values via the `exog` argument.
```

**Review by yarnabrina — APPROVED** [review]
Looks good to me. Approved subject to reviews of @ciaran-g / @fkiraly regarding behaviour of `self._X` and its interaction with X y same index tag.

File: sktime/forecasting/base/adapters/_statsmodels.py
Comment (Abhay-Lejith) [reply sktime/forecasting/base/adapters/_statsmodels.py]: @yarnabrina so do we mark this conversation as resolved or are we waiting for @fkiraly /@ciaran-g to review it as well

Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_statsmodels.py]: I'd wait for at least 2 days. Otherwise, we can go ahead and merge.

Comment (ciaran-g) [reply sktime/forecasting/base/adapters/_statsmodels.py]: Not 100% sure on the tag behaviour myself unfortunately, but the changes look ok and sensible to me 👍

**Pull request merged**
